[SR-331] Generate fix-its for confusable characters

amackworth opened the pull request:
Addresses issue SR-331 by adding warning fix-its for possible Unicode confusions. This check only runs after either the lexer encounters an invalid character or an identifier fails to resolve. (All tests pass with `utils/build-script -RT` on OS X 10.11.)

A few possible issues that I'd love feedback on:
- This patch includes changes to both the lexer and type checker. Should I break it into two different commits?
  - Relatedly, I had to expose two UTF-8 helper functions from `Lexer.cpp` such that `TypeCheckConstraints.ccp` could use them. Should these be copied into the latter instead or moved to a shared location?
  - Additionally, I had to duplicate the warning message across both `DiagnosticsParse.def` and `DiagnosticsSema.def`. Should these also be extracted to another shared location?
- I created a new test file as I wasn't sure that any of the existing ones were close enough in topic.
- These checks rely on some slightly hacky macros. Would there be a better way to refactor?

Finally, given this is my first contribution, I welcome any guidance to match the style of the rest of the project!

amackworth committed:
[SR-331] Generate fix-its for confusable characters.

gribozavr commented on include/swift/Parse/Confusables.def:
```diff
@@ -0,0 +1,119 @@
+#define CONFUSABLE(confused, expected)
+#endif
+
+CONFUSABLE(0x2010, 0x2d)
```
It would help to describe the procedure you used to generate this table, so that it can be updated for a new version of Unicode when it comes out.

Please also mention in comments the Unicode version you used this time.

dmishe commented on include/swift/Parse/Confusables.def:
```diff
@@ -0,0 +1,119 @@
+
+CONFUSABLE(0x2010, 0x2d)
+CONFUSABLE(0x2011, 0x2d)
+CONFUSABLE(0x2012, 0x2d)
```
It might help to put a comment next to each of actual characters being replaced, like `// a > b`, especially if this is auto-generated.

amackworth commented:
@gribozavr: Just added the version/source to the table.

Generating this table was a bit of a messy process, mainly involving a quick Python script that was given the list of punctuation from `Tokens.def` plus a few other basic operators and went through the list from http://www.unicode.org/Public/security/8.0.0/confusables.txt, outputting that list of pairs. I also had to manually trim the list to avoid collisions.

I could also post the script here if you'd like, but it would take quite a bit of cleanup. :)

gribozavr commented:
@amackworth We usually have all those tools checked in as a part of the repository, and, where feasible, they run as a part of the build.  Check out `./lib/Basic/UnicodeExtendedGraphemeClusters.cpp.gyb` for example.

In this case, I think it is feasible to run it as a part of the build if you turn the header into a function in a .cpp file that takes a unicode scalar and returns you a replacement or 0 if it is not confusable.  This will also improve code size.

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -436,6 +436,9 @@ ERROR(unspaced_unary_operator,sema_nb,none,
 
 ERROR(use_unresolved_identifier,sema_nb,none,
       "use of unresolved %select{identifier|operator}1 %0", (Identifier, bool))
+WARNING(confusable_character,sema_nb,none,
```
Rather than include this in two different Diagnostics files, please move it to DiagnosticsCommon.def.

jrose-apple replied:
Oh, and it should also be a NOTE rather than a WARNING because it should be attached to the previous error.

jrose-apple commented on include/swift/Parse/Lexer.h:
```diff
@@ -24,6 +24,8 @@
 
 namespace swift {
+  uint32_t validateUTF8CharacterAndAdvance(const char *&Ptr, const char *End);
+  bool EncodeToUTF8(unsigned CharValue, SmallVectorImpl<char> &Result);
```
If you're going to publish these—and I'm not convinced that's a good idea yet, see below—please give them proper doc comments, and change the second to `encodeToUTF8` to follow dominant naming conventions.

amackworth replied:
Sorry, I'm not sure what you're referring to below? I agree that I'd prefer not to publish them, but unfortunately I currently need `validateUTF8CharacterAndAdvance()` to iterate through the identifier in `TypeCheckConstraints.cpp`.

jrose-apple commented on lib/Parse/Lexer.cpp:
```diff
@@ -1477,6 +1477,29 @@ void Lexer::lexImpl() {
+          EncodeToUTF8(codepoint, buffer);
+          confusedChar = buffer.str();
+        } else {
+          confusedChar = StringRef(std::string(1, (char)codepoint));
```
StringRef is a non-owning type, so both branches here result in references to deallocated stack memory. We generally think it's okay to make the diagnostic path slower, so just taking the slow path all the time here is okay rather than trying to be clever.

amackworth replied:
Agreed, but `EncodeToUTF8` asserts that codepoint has to be greater than `0x80`, and I wasn't sure that I should change the contract of the function.

amackworth replied:
Hmm...I'm a little confused as to what this means. Are you saying I should use `EncodeToUTF8()` regardless of whether the codepoint is greater than `0x80`? If so, that would mean changing the contractual semantics of the helper function, which worries me.

gribozavr replied:
`std::string` constructs a temporary string.  `StringRef` creates a reference to that temporary object.  The object goes away at the end of this statement, and `confusedChar` refers to freed memory when it is used below.

amackworth replied:
Oh right, meant to fix that! I'll take a look tomorrow.

jrose-apple commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -451,6 +451,37 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
       diagnose(Loc, diag::use_unresolved_identifier, Name,
                UDRE->getName().isOperator())
         .highlight(Loc);
+      const char *buffer = Name.get();
```
I think we do want a similar error here, but
(a) the code duplication makes me unhappy, and
(b) if someone actually does define an unusual operator with multiple characters in it, being warned about each of them individually seems suboptimal.

Maybe a better answer would be to "deconfuse" the entire operator identifier and tell the user to try that, with special-case wording for single-character operators. What do you think?

amackworth replied:
Yeah, that sounds much better!

jrose-apple commented:
For bonus points, it would be awesome™ to recover as if the user had typed the operator in question.

amackworth commented:
Following @gribozavr's suggestion, the table of confusable characters is now generated by `gyb` as part of the build process. (I wasn't sure where to put the `confusables.txt` file, so it's living in `lib/Parse` for now.)

Additionally, as @jrose-apple proposed, the fix-it now "deconfuses" the entire identifier at once, rather than generating errors for each individual possibly-confused character.

Thank you for the feedback!

gribozavr commented:
@amackworth Thanks! `utils/UnicodeData` is the place for Unicode data files.

gribozavr commented on lib/Parse/Confusables.cpp.gyb:
```diff
@@ -0,0 +1,64 @@
+    noPrefix = hexString[2:]
+    modifiedHex.append((((4 - len(noPrefix)) * "0") + noPrefix).upper())
+
+f = open('confusables.txt', 'r')
```
`with open('confusables.txt', 'r') as f:` ?

gribozavr commented on include/swift/Parse/Confusables.h:
```diff
@@ -0,0 +1,23 @@
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
```
Please add the final line:

```
//===----------------------------------------------------------------------===//
```

gribozavr commented on include/swift/Parse/Confusables.h:
```diff
@@ -0,0 +1,23 @@
+
+namespace swift {
+namespace confusable {
+  uint32_t characterIsConfusable(uint32_t codepoint);
```
Using `Is` in the name suggests that the function returns `bool`, which it does not.  I would suggest something like `tryConvertConfusableCharacterToASCII`.

Also, a doc comment wouldn't hurt (specifically documenting what the return value is if the character is not confusable).

amackworth commented:
@gribozavr: Done! :smile:

gribozavr commented on include/swift/Parse/Confusables.h:
```diff
@@ -0,0 +1,27 @@
+namespace confusable {
+  /// Given a UTF-8 codepoint, determines whether it appears on the Unicode
+  /// specification table of confusable characters and maps to punctuation,
+  /// and either returns either the expected ASCII character or ~0U.
```
Why not use `0` as the "not confusable" marker?  It would make call site simpler:

```
if (uint32_t replacement = tryConvertConfusableCharacterToASCII(c)) { ...
```

amackworth replied:
Ah, right! Makes sense.

amackworth replied:
Fixed!

amackworth commented:
So, I just fixed the `StringRef` issue, in addition to realizing that I didn't actually need to export `EncodeToUTF8`! I also removed the `assert` at the beginning of said helper to avoid branching, but I'm more than a little worried about that, given that it changes the contractual semantics of the function. Finally, I added a doc string for `validateUTF8CharacterAndAdvance`.

I'm still working on possibly recovering as if the user had typed in the expected expression as @jrose-apple suggested, but I'm not sure how to restart the name lookup process since the `Identifier` is immutable from the perspective of `swiftSema`.

(cc @gribozavr)

amackworth commented:
Just checking in post-holiday on the status of this PR, and if there's anything else you'd like me to change! In particular, I'd really appreciate any suggestions for how to recover from the warning. (cc @gribozavr and @jrose-apple)

lattner commented:
hi @amackworth, @gribozavr is out on vacation this week, but will be back next week.

One comment from me: please do not use gyb for C++ source files.  It would be better to use the C preprocessor directly, probably with a ".def" style approach like we do for other things in swift/include.

gribozavr commented:
@lattner The reason to use gyb in this case is that it allows us to construct the C++ source directly from tables published with the Unicode spec.  The .def file approach requires us to check in a separate file derived from the vanilla tables.

jrose-apple commented:
Sorry, haven't had time to look at this properly, but still planning to give you an answer about recovery.

lattner commented:
@gribozavr Ok, that's a nice win, but seriously, please do not use gyb for c++ files.

jrose-apple commented:
@lattner, would an intermediate step of .gyb to .def be good enough? Our build system already supports .gyb, and having to manually update a .def file, even with a script checked into utils/, seems like an unnecessary extra step.

jrose-apple commented on lib/Parse/Lexer.cpp:
```diff
@@ -44,11 +45,7 @@ using clang::isWhitespace;
-static bool EncodeToUTF8(unsigned CharValue,
-                         SmallVectorImpl<char> &Result) {
-  assert(CharValue >= 0x80 && "Single-byte encoding should be already handled");
+bool EncodeToUTF8(unsigned CharValue, SmallVectorImpl<char> &Result) {
```
This isn't used outside the file anymore, so you can leave it alone.

jrose-apple commented on lib/Parse/Lexer.cpp:
```diff
@@ -1477,6 +1471,22 @@ void Lexer::lexImpl() {
+        uint32_t expectedCodepoint;
+        if ((expectedCodepoint =
+            confusable::tryConvertConfusableCharacterToASCII(codepoint))) {
+          std::string expected = std::string(1, (char)expectedCodepoint);
```
Please use a SmallString here too (which implicitly converts to StringRef), so that we don't have any heap allocation.

jrose-apple commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -451,6 +452,36 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
         .highlight(Loc);
+
+      const char *buffer = Name.get();
+      std::string expectedIdentifier = std::string();
```
Now that this is at the top level, you can use a SmallString instead of std::string here too.

jrose-apple commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -27,6 +27,7 @@
 #include "swift/AST/TypeCheckerDebugConsumer.h"
 #include "swift/Basic/Fallthrough.h"
 #include "swift/Parse/Lexer.h"
+#include "swift/Parse/Confusables.h"
```
Please keep the includes sorted.

jrose-apple commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -451,6 +452,36 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
+        offset += length;
+      }
+      
+      if (isConfused) {
```
So, to recover I would call `lookupUnqualified` again here and see if you get results back. If you do get results, you can diagnose this as a single `ERROR` with a fix-it and then continue in this function. If you *don't* get any results, I'm not sure we should even say it's confused; using the ASCII characters won't improve anything.

(Well, maybe it will because maybe you've confused characters *and* juxtaposed operators, but that seems kind of unlikely.)

jrose-apple commented on lib/Parse/Lexer.cpp:
```diff
@@ -1477,6 +1471,22 @@ void Lexer::lexImpl() {
+            .fixItReplaceChars(getSourceLoc(CurPtr-1),
+                               getSourceLoc(tmp),
+                               expectedChar);
+        }
```
Recovering here would mean jumping to the top of the switch with the expected character instead, but I'm not sure we're set up to handle that very well. We probably assume all over the place that all the ASCII characters are only one byte wide. I guess it's okay not to do anything here.

jrose-apple commented on lib/Parse/Confusables.cpp.gyb:
```diff
@@ -0,0 +1,63 @@
+
+#include "swift/Parse/Confusables.h"
+
+uint32_t swift::confusable::tryConvertConfusableCharacterToASCII(uint32_t codepoint) {
```
If this really only supports ASCII, why return `uint32_t` instead of `char`?

lattner commented:
> @lattner, would an intermediate step of .gyb to .def be good enough? 

Yes, I think the best thing in this case is to have a python (or whatever) script in swift/utils that is manually run to produce the .def file.  The .def file would be checked into the tree.

dabrahams commented:
on Sun Jan 17 2016, Chris Lattner <notifications-AT-github.com> wrote:

>> @lattner, would an intermediate step of .gyb to .def be good enough? 
>
> Yes, I think the best thing in this case is to have a python (or
> whatever) script in swift/utils that is manually run to produce the
> .def file.  The .def file would be checked into the tree.

Why is this better than just using gyb?

lattner commented:
Because gyb is a necessary evil used in the stdlib.  We want to eliminate its use over time, not spread it to other parts of the code base.

More broadly, we have a solution to this sort of problem established in the LLVM community, and we should use that solution.

jrose-apple commented:
We do have a solution, but that solution is TableGen. Not everything fits in a .def file. I'd argue that gyb is better than TableGen.

lattner commented:
Tablegen isn't the only approach.  LLVM uses perfect shuffle and other utils that generate a .cpp or .h file that is checked into the tree.  Take that approach (but write the script in python if that is what you want) I agree that writing a tblgen backend is the wrong way to go.

jrose-apple commented:
Why is "run a script manually whenever something changes" better than "run a script as part of the build whenever something changes"?

lattner commented:
There is virtue in keeping the build machinery (e.g. cmake goop) as simple as possible, and making builds run fast.  I'm not concerned about this in terms of build time, but it is a slippery slope that we should definitely not slide down.

gribozavr commented:
We already have all the CMake code that supports C++ and gyb, it is required anyway for Swift code that uses gyb.  There's nothing to simplify in CMake if C++ code didn't use gyb.

dabrahams commented:
on Tue Jan 19 2016, Dmitri Gribenko <notifications-AT-github.com> wrote:

> We already have all the CMake code that supports C++ and gyb, it is
> required anyway for Swift code that uses gyb.  There's nothing to
> simplify in CMake if C++ code didn't use gyb.

Sometimes gyb is the right tool for the job.  If switching away from it
is going to create maintenance headaches, require infrastructure work,
or make development error-prone, I don't understand why we would switch.

lattner commented:
I completely understand that the cmake goop already exists.  That doesn't address the slippery slope, nor does it address the desire to make gyb go away entirely.

Further, these tables do not change frequently.  I do not see a reason to regenerate the output as part of the build process.
